### ISSUE TITLE
Add wasCompiledWithMPI to SolverInterface

### ISSUE THREAD

**Issue by BenjaminRueth** [precice/precice#311]
We should provide an API function to be able to find out whether preCICE was compiled with MPI or not. This issue is related to #300 

**Why do we need this functionality?** As an example, the python interface of preCICE requires you to `import mpi4py`, if preCICE has been compiled with MPI (for example https://github.com/precice/precice/pull/299). If preCICE has been compiled without MPI, you do not need this import statement. 

Putting `import mpi4py` everywhere is the safe way to solve this problem. However, this also adds an unnecessary dependency for users that did not compile preCICE with MPI. Ideally, we can use the following lines in the future:

in `__init__.py` (is this the correct solution? See https://github.com/precice/precice/issues/300#issue-415085361)
```
if precice.was_compiled_with_MPI():
    import mpi4py
```

in `setup.py`
```
dependencies = [...]

if precice.was_compiled_with_MPI():
    dependencies.append = ["mpi4py"]

setup(...
      install_requires=dependencies,
      ...)

**Comment by BenjaminRueth** [comment]
For instructions, how to reproduce the error, see https://github.com/precice/precice/pull/299#issuecomment-469421697.

**Comment by uekerman** [comment]
I am not to sure whether we need this. Please keep in mind that maybe 1% of our users will compile preCICE without MPI. This is only needed if you have a closed source code that comes with a pre-compiled MPI which interferes with the MPI dependency of preCICE. I am not even sure if anybody uses this at the moment. And then this closed-source code needs to use the Python bindings. Adding `mpi4py` by default seems to be the easier solution. Any objections to this though? I vote for making this issue a `wontfix`.

**Comment by floli** [comment]
A possible implementation in `SolverInterface.hpp` is just like
```
bool SolverInterfaceImp::wasCompiledWithMPI 
{
    return not #PRECICE_NO_MPI
}
```
true? So next to no effort? @BenjaminRueth stated a use case for that and I concur with him. Especially on weird, but unfortunately not too uncommon systems (windows & proprietary solvers) installing a python library which in turn requires MPI might be some work.

**Comment by uekerman** [comment]
Well, you also need to port this to all bindings, to write tests for it, write documentation ...
and I don't see the use case. Windows & proprietary solvers will not need the Python bindings. Let's please only do this if we really need it.

**Comment by BenjaminRueth** [comment]
This feature would help us resolving #312. As a compromise we can also only add it to the Python bindings for now and make it a `wontfix` for the remaining bindings. 

@uekerman do you agree? If not, we can try to find another way for resolving #312 and close this issue as `wontfix`.

**Comment by MakisH** [comment]
I think it would be quite useful to have this information, but maybe we don't need to get this information from a function call.

Can we get this piece of information from CMake somehow? We only need it at build time, not at runtime.

**Comment by uekerman** [comment]
> As a compromise we can also only add it to the Python bindings

This is no good solution in my mind. We now worked hard to make all bindings consistent. 

> Can we get this piece of information from CMake somehow? We only need it at build time, not at runtime.

If that's possible that would be great

**Comment by fsimonis** [comment]
> Can we get this piece of information from CMake somehow? We only need it at build time, not at runtime.

Yes we can, but the installation of the bindings is now completely decoupled from the library.
We could even put the python bindings into another repository and publish them via PyPI so that users could just do a `pip install -U precice`.

**Comment by MakisH** [comment]
Yes, but we always install some information into the `[prefix]/lib/cmake/precice` directory. Could we get something like `COMPILED_WITH_MPI` in there, which could then be parsed by the bindings?

Or any other constant stored together with the library and read at compile time. Am I missing something here?

**Comment by fsimonis** [comment]
@MakisH  Yes, We could put the configuration information into the the `preciceConfig.cmake` which would also expose the configuration to the user. This would be generally useful.
However, we would have to try and locate this config, which sounds like a nightmare to me.

As an easy workaround we could simply initialize MPI if it is available on the system.
This way, the user has to simply install mpi4py if he needs it and it will loaded automatically.

In the package `__init__` we could write this:
```python
try:
    from mpi4py import MPI  # Initialize MPI
except ImportError:
    pass
```

Problem is that doing it this way, the system packages will always require an additional install `mpi4py`. We could simply document that though, or print some information while installing.

**Comment by MakisH** [comment]
> Problem is that doing it this way, the system packages will always require an additional install `mpi4py`.

I am not sure I understand this. What do you mean "system packages" here? Installing mpi4py without needing it is what we want to avoid.

I would rather rename this issue to "Python bindings: don't import mpi4py if not needed", close it with `wont-fix` for now and reopen it in the future, if anyone faces this problem.

**Comment by fsimonis** [comment]
@MakisH With "system packages" I mean the current debian package and possible future distro packages. We currently build these packages with MPI and PETSc enabled.

Reliably detecting whether preCICE was built with MPI will be a pain to get right and maintain.

**Comment by uekerman** [comment]
Let's not waste too much time here. We have two options:

(1) Put in the bindings:
```
try:
    from mpi4py import MPI  # Initialize MPI
except ImportError:
    pass
```
with a decent documentation.
 
(2) Keep it as it was: The user has to import MPI if he uses preCICE with MPI. Also no big issue.

**Comment by BenjaminRueth** [comment]
If we implement #261, then issue will be solved automatically. `getVersion()` is supposed to provide information on build configuration (like MPI), as well.

**Comment by uekerman** [comment]
Can be closed right?

**Comment by BenjaminRueth** [comment]
> 
> 
> Can be closed right?

I would prefer to wait for #261 to see whether this actually resolves the original problem: Can we get information about MPI from within the python bindings?
